### ISSUE TITLE
Intermediate variable is needed for construct-typing

### ISSUE THREAD

**Issue by timrid** [microsoft/pyright#5404]
**Describe the bug**
I am the author of the [construct-typing](https://github.com/timrid/construct-typing) package. In it I created stubs for the [construct](https://github.com/construct/construct) library. Until version v1.1.309 this works fine. But since v1.1.310 until now with v1.1.316 it produces several errors. One of them is described below.

**To Reproduce**
`pip install construct-typing==0.5.6`

```python
import construct as cs

# working
intermediate = cs.Array(5, cs.Byte)
fmt_working = cs.GreedyRange(intermediate)
reveal_type(fmt_working)  # information: Type of "fmt_working" is "GreedyRange[ListContainer[int], List[int], ListContainer[ListContainer[int]], List[List[int]]]"

# not working
fmt_not_working = cs.GreedyRange(cs.Array(5, cs.Byte))
reveal_type(fmt_not_working)  # information: Type of "fmt_not_working" is "GreedyRange[ParsedType@Array, BuildTypes@Array, Unknown, Unknown]"
```

Output of `pyright example_pyright.py`:

```
C:\...\example_pyright.py
  C:\...\example_pyright.py:6:13 - information: Type of "fmt_working" is "GreedyRange[ListContainer[int], List[int], ListContainer[ListContainer[int]], List[List[int]]]"
  C:\...\example_pyright.py:9:46 - error: Argument of type "FormatField[int, int]" cannot be assigned to parameter "subcon" of type "Construct[SubconParsedType@Array, SubconBuildTypes@Array]" in function "__new__"
    "FormatField[int, int]" is incompatible with "Construct[int, SubconBuildTypes@Array]"
      TypeVar "BuildTypes@Construct" is contravariant
        "object*" is incompatible with "int" (reportGeneralTypeIssues)
  C:\...\example_pyright.py:9:1 - error: Type of "fmt_not_working" is partially unknown
    Type of "fmt_not_working" is "GreedyRange[ParsedType@Array, BuildTypes@Array, Unknown, Unknown]" (reportUnknownVariableType)
  C:\...\example_pyright.py:10:13 - information: Type of "fmt_not_working" is "GreedyRange[ParsedType@Array, BuildTypes@Array, Unknown, Unknown]"
2 errors, 0 warnings, 2 informations
```

**Expected behavior**
The intermediate variable should not be required.

**VS Code extension or command-line**
command line, pyright 1.1.316

**Comment by erictraut** [comment]
This issue stems from the way `Array.__new__` is defined. `Array` is a generic class that is parameterized with four type parameters. If I explicitly specialize `Array` and then call the constructor `Array[int, int, int, int](<constructor args>)`, I would expect to receive an instance of type `Array[int, int, int, int]`. However, the `__new__` method modifies the type arguments and returns an instance of `Array[int, int, ListContainer[int], list[int]]`. Is this intended? This looks like a bug to me. The `__new__` method should honor the type arguments of a specialized class, not mutate them.

```python
import construct as cs
a = cs.Array[int, int, int, int](5, cs.Byte)
reveal_type(a) # Array[int, int, ListContainer[int], list[int]]
```

This is interfering with pyright's ability to perform bidirectional type inference, which is used in the "not working" example but is not used in the "working" example. That explains the difference in behavior that you're seeing.

Prior to 1.1.309, pyright ignored the return type of `__new__` if there was an `__init__` method. But in 1.1.310, I rewrote the constructor logic so it more closely matches the runtime behavior. Pyright now honors the return type annotation of `__new__` even if there is an `__init__` method. Mypy always ignores the annotated return type of `__new__`, but there is an open bug report on this topic. If mypy were to fix this bug, you would likely see the same behavior in mypy.

Another odd thing about the declaration for classes `Array` and `GreedyRange` is that the stubs declare both a `__new__` and an `__init__` method. Typically stubs provide one or the other but not both. It's OK to provide both, but in this case, it forces the type checker to perform redundant argument checks. I'm guessing that the reason you added both was to work around the differences between mypy and pyright. Is that correct?

And another odd thing is that the `__init__` method in both `Array` and `GreedyRange` have an annotated `self` parameter that uses TypeVars. Affecting the constructed type via an annotated `self` within an `__init__` method is a hacky mechanism that mypy introduced (unfortunately). This mechanism was never properly spec'ed, and it's dependent on mypy's internal implementation details. I initially refused to add support for this mechanism in pyright because it's not part of the typing standard, but then typeshed stubs started to rely on it for some stdlb classes. We came to an agreement that its meaning was well-defined in cases where the annotated type didn't include type variables, so pyright now supports that. However, you're using type variables in the `self` annotation. Pyright will therefore not honor that annotation.

**Comment by erictraut** [comment]
I added a workaround in pyright to better handle this case. However, I recommend that you look at the issue I mentioned above because this looks to me like a bug in the constructor definition.

**Comment by timrid** [comment]
Thanks for the detailed analysis and answer. 

You guess right that I defined `__new__` and `__init__` to be compatible with both pyright and mypy. This was not my preferred choice, but when I wrote this it was the only way to make it work with both.

I was aware that you are not allowed to annotate `self` in pyright. You had noted that 2 years ago (https://github.com/python/typeshed/issues/4846#issuecomment-753435499) and it is now also describte in the typing-docs [here](https://typing.readthedocs.io/en/latest/source/stubs.html#classes). But for mypy it was necessary for it to work.

I started rebuilding all the type stubs to eliminate the `__new__` methods. In most places this now works with mypy and pyright. I changed the code from this:

```python
class Array(
    Subconstruct[
        SubconParsedType,
        SubconBuildTypes,
        ParsedType,
        BuildTypes,
    ]
):
    count: ConstantOrContextLambda[int]
    discard: bool
    def __new__(
        cls,
        count: ConstantOrContextLambda[int],
        subcon: Construct[SubconParsedType, SubconBuildTypes],
        discard: bool = ...,
    ) -> Array[
        SubconParsedType,
        SubconBuildTypes,
        ListContainer[SubconParsedType],
        t.List[SubconBuildTypes],
    ]: ...
    def __init__(
        self: Array[
            SubconParsedType,
            SubconBuildTypes,
            ListContainer[SubconParsedType],
            t.List[SubconBuildTypes],
        ],
        count: ConstantOrContextLambda[int],
        subcon: Construct[SubconParsedType, SubconBuildTypes],
        discard: bool = ...,
    ) -> None: ...
```

to this:

```python
class Array(
    Subconstruct[
        SubconParsedType,
        SubconBuildTypes,
        ListContainer[SubconParsedType],  # type: ignore
        t.List[SubconBuildTypes],  # type: ignore
    ]
):
    count: ConstantOrContextLambda[int]
    discard: bool
    def __init__(
        self,
        count: ConstantOrContextLambda[int],
        subcon: Construct[SubconParsedType, SubconBuildTypes],
        discard: bool = ...,
    ) -> None: ...
```

I only run into problems when different generics need to be used based on the constructor type hints. Here I have to use the `__new__` method and adjust the return type based on the parameter types. But then the same probem you pointed out occurs again. Here is a small self contained example of what I mean. Do you have any idea how to make this better?

```python
import typing as t

T = t.TypeVar("T")

ParsedType = t.TypeVar("ParsedType", covariant=True)
BuildTypes = t.TypeVar("BuildTypes", contravariant=True)

class Construct(t.Generic[ParsedType, BuildTypes]):
    def parse(self, data: bytes) -> ParsedType: ...
    def build(self, obj: BuildTypes) -> bytes: ...

Int8ub: Construct[int, int]

class Const(Construct[ParsedType, BuildTypes]):
    value: BuildTypes
    @t.overload
    def __new__(
        cls,
        value: bytes,
    ) -> "Const[bytes, t.Optional[bytes]]": ...
    @t.overload
    def __new__(
        cls,
        value: BuildTypes,
        subcon: Construct[ParsedType, BuildTypes],
    ) -> "Const[ParsedType, t.Optional[BuildTypes]]": ...

a = Const(5, Int8ub)
reveal_type(a)  # Const[int, Literal[5] | None]

b1 = Const(b"1234")
reveal_type(b1)  # Const[bytes, bytes | None]

b2 = Const[int, int](b"1234")  # <---- explicit specialization of `Const`
reveal_type(b2)  # Const[bytes, bytes | None]  <---- same problem as you described 
```

**Comment by erictraut** [comment]
If you use class-scoped type variables in your constructor's parameters, then the constructor should return `Self`. If `__new__` does not return `Self`, then the constructor should not use class-scoped TypeVars in its signature. It can use method-scoped type variables. Many classes in the typeshed stubs do this.

Here's an attempt to address this problem. Note that I've explicitly annotated the `cls` variable in the `__new__` methods to enforce that a specialized `cls` type must conform to the constructed type. I've also switched from class-scoped TypeVars to method-scoped TypeVars in the `__new__` method.

```python
T0 = TypeVar("T0", contravariant=True)
T1 = TypeVar("T1", contravariant=True)

class Const(Construct[ParsedType, BuildTypes]):
    value: BuildTypes

    @overload
    def __new__(
        cls: "type[Const[bytes, bytes | None]]",
        value: bytes,
    ) -> "Const[bytes, bytes | None]":
        ...

    @overload
    def __new__(
        cls: "type[Const[T0, T1 | None]]",
        value: T1,
        subcon: Construct[T0, T1],
    ) -> "Const[T0, T1 | None]":
        ...

b2 = Const[int, int](b"1234")  # This now generates an error in pyright as it should (but strangely, not in mypy)
reveal_type(b2)  # Never

b3 = Const[bytes, bytes](b"1234") # This also generates an error

b4 = Const[bytes, bytes | None](b"1234") # This does not generate an error, which is correct

b5 = Const[int, int](4, Int8ub) # This generates an error

b6 = Const[int, int | None](4, Int8ub) # This does not, which is correct
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
